src/common/src/array/list_array.rs

```diff
@@ -431,118 +376,189 @@ impl ListValue {
         let bytes = serde_bytes::ByteBuf::deserialize(deserializer)?;
         let mut inner_deserializer = memcomparable::Deserializer::new(bytes.as_slice());
-        let mut values = Vec::new();
+        let mut builder = datatype.create_array_builder(0);
         while inner_deserializer.has_remaining() {
-            values.push(memcmp_encoding::deserialize_datum_in_composite(
+            builder.append(memcmp_encoding::deserialize_datum_in_composite(
                 datatype,
                 &mut inner_deserializer,
             )?)
         }
-        Ok(Self::new(values))
+        Ok(Self::new(builder.finish()))
+    }
+
+    // Used to display ListValue in explain for better readibilty.
+    pub fn display_for_explain(&self) -> String {
+        // Example of ListValue display: ARRAY[1, 2, null]
+        format!(
+            "ARRAY[{}]",
+            self.iter()
+                .map(|v| {
+                    match v.as_ref() {
+                        None => "null".into(),
+                        Some(scalar) => scalar.to_text(),
+                    }
+                })
+                .format(", ")
+        )
+    }
+
+    /// Returns a mutable slice if the list is of type `int64[]`.
+    pub fn as_i64_mut_slice(&mut self) -> Option<&mut [i64]> {
+        match self.values.as_mut() {
+            ArrayImpl::Int64(array) => Some(array.as_mut_slice()),
+            _ => None,
+        }
+    }
+}
+
+impl PartialOrd for ListValue {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for ListValue {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.as_scalar_ref().cmp(&other.as_scalar_ref())
+    }
+}
+
+impl<T: PrimitiveArrayItemType> FromIterator<Option<T>> for ListValue {
+    fn from_iter<I: IntoIterator<Item = Option<T>>>(iter: I) -> Self {
+        Self::new(iter.into_iter().collect::<PrimitiveArray<T>>().into())
+    }
+}
+
+impl<T: PrimitiveArrayItemType> FromIterator<T> for ListValue {
+    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
+        Self::new(iter.into_iter().collect::<PrimitiveArray<T>>().into())
+    }
+}
+
+impl FromIterator<bool> for ListValue {
+    fn from_iter<I: IntoIterator<Item = bool>>(iter: I) -> Self {
+        Self::new(iter.into_iter().collect::<BoolArray>().into())
+    }
+}
+
+impl<'a> FromIterator<Option<&'a str>> for ListValue {
+    fn from_iter<I: IntoIterator<Item = Option<&'a str>>>(iter: I) -> Self {
+        Self::new(iter.into_iter().collect::<Utf8Array>().into())
+    }
+}
+
+impl<'a> FromIterator<&'a str> for ListValue {
+    fn from_iter<I: IntoIterator<Item = &'a str>>(iter: I) -> Self {
+        Self::new(iter.into_iter().collect::<Utf8Array>().into())
+    }
+}
+
+impl FromIterator<ListValue> for ListValue {
+    fn from_iter<I: IntoIterator<Item = ListValue>>(iter: I) -> Self {
+        Self::new(iter.into_iter().collect::<ListArray>().into())
+    }
+}
```

Comment:
Will this be possibly misused?

Reply:
Humm, I think it's clear that collecting a list of lists produces a list of lists. 😇